core/sdk-core/src/main/java/software/amazon/awssdk/core/progress/listener/ProgressListener.java

```diff
@@ -0,0 +1,395 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.progress.listener;
+
+import software.amazon.awssdk.annotations.Immutable;
+import software.amazon.awssdk.annotations.SdkPreviewApi;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.annotations.ThreadSafe;
+import software.amazon.awssdk.core.progress.snapshot.ProgressSnapshot;
+import software.amazon.awssdk.http.SdkHttpRequest;
+
+/**
+ * The {@link ProgressListener} interface may be implemented by your application in order to receive event-driven updates on
+ * the progress of a service call. When you construct an {@link PutObjectRequest} or {@link
+ * UploadPartRequest} request submitted to the Sdk, you may provide a variable number of {@link
+ * ProgressListener}s to be associated with that request. Then, throughout the lifecycle of the request,
+ * the Sdk
+ * will invoke the provided {@link ProgressListener}s when important events occur, like additional bytes being transferred,
+ * allowing you to monitor the ongoing progress of the transaction.
+ * <p>
+ * Each {@link ProgressListener} callback is invoked with an immutable {@link Context} object. Depending on the current
+ * lifecycle
+ * of the request, different {@link Context} objects have different attributes available (indicated by the provided context
+ * interface). Most notably, every callback is given access to the current {@link ProgressSnapshot}, which contains
+ * helpful progress-related methods like {@link ProgressSnapshot#transferredBytes()} and {@link
+ * ProgressSnapshot#ratioTransferred()}.
+ * <p>
+ * A successful transfer callback lifecycle is sequenced as follows:
+ * <ol>
+ *     <li>{@link #requestPrepared(Context.RequestPrepared)} - A new Request has been initiated. This method is called
+ *     exactly once per transfer.</li>
+ *     <ul>Available context attributes:
+ *         <li>{@link Context.RequestPrepared#request()}</li>
+ *         <li>{@link Context.RequestPrepared#progressSnapshot()}</li>
+ *     </ul>
+ *     <li>{@link #requestBytesSent(Context.RequestBytesSent)} - Additional bytes have been sent. This
+ *     method may be called many times per request, depending on the request payload size and I/O buffer sizes.
+ *     <li>{@link #responseBytesReceived(Context.ResponseBytesReceived)} - Additional bytes have been received. This
+ *     method may be called many times per request, depending on the response payload size and I/O buffer sizes.
+ *     <li>{@link #executionSuccess(Context.ExecutionSuccess)} - The transfer has completed successfully. This method is called
+ *     exactly once for a successful transfer.</li>
+ * </ol>
+ * For every failed attempt {@link #attemptFailure(Context.AttemptFailure)}  will be called exactly once.
```

Comment:
I'm worried people might be hung up on "exactly once" and think that it only gets called once per execution (even though that's not what this actually says). Do we need the "exactly once" here or is "For every failed attempt, {@link #attemptFailure(Context.AttemptFailure)} will be called." sufficient?

Reply:
RIght. I am making the language tighter